Chapter15: Working with SpriteFonts

AristurtleDev opened the pull request:
Live Preview: https://aristurtle.net/tutorials/building_2d_games/15_working_with_spritefonts/

AristurtleDev committed:
Initial chapter 15 draft

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,229 @@
+2. Data that tracks the position, size, and spacing of each character.
+3. Kerning information for adjusting spacing between specific character pairs.
+
+The texture atlas approach means fonts are rendered as sprites, using the same [**SpriteBatch**](xref:Microsoft.Xna.Framework.Graphics.SpriteBatch) system you learned about for drawing textures.
```
Probably worth a note as to why it is a texture atlas and not just a "Font", as installed fonts cannot be guaranteed on all platforms and graphics cards do not understand fonts.

AristurtleDev replied:
Updated with a list of reasons why MonoGame uses texture atlas for fonts including not relying on system installed fonts, consistency of rendered glyphs across platforms, and performance (no texture swapping when rendering log strings).

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,229 @@
+
+#### FontName
+
+The `<FontName>` element specifies which font to use. By default, it references "Arial".  When a font name is specified just by name like this, it is required that the font be installed on the system where the content is built.
```
Add a note that MonoGame recommends changing the Arial font if you are targeting any platforms other than Windows.  Arial is a legacy of XBA and only included in Windows builds.

Maybe add a recommendation for a Font, MonoGame now uses Roboto from google fonts by default.

AristurtleDev replied:
Added an Important admonition block explaining that Arial is only available on windows builds and is default default due to legacy XNA.  Included recommendation for using Roboto as an alternative.

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,229 @@
+
+#### Size
+
+The `<Size>` element controls the font size in points. For pixel fonts, keeping this at a small value preserves the crisp pixel look. For smoother, anti-aliased fonts, you might want a larger value.
```
Worth a larger note about size since there are Sooooo many patterns, considerations to include:

* Resolution your game is running at - small fonts on a large rez look awful
* You may want to include several font files for different text resolutions, or consider scaling down the text at draw time to meet the resolution needs of your game

It is a fine balance and no one size fits all :D

AristurtleDev replied:
Now includes information on different factors to consider when choosing a font size and different approaches to using multiple font sizes, such as multiple SpriteFont Descriptions or a single SpriteFont Description of a large font and the tradeoffs between each

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,229 @@
+
+#### UseKerning
+
+The `<UseKerning>` element determines whether to use kerning information from the font. Kerning adjusts the spacing between specific pairs of characters for more visually pleasing results. For most fonts, you'll want to leave this as `true`.
```
Although some font files do not react well to Kerning, like Arial for instance :P

AristurtleDev replied:
Added an additional note about kerning and to disable it as a troubleshooting step if spacing of characters is off.

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,229 @@
+
+Let's start by understanding how text rendering works in MonoGame.
+
+## Understanding SpriteFonts
```
(Optional) Could also possibly do with a small section on "choosing a font" and the considerations you need to consider:

* Size
* Scale
* License
* Format
* etc

AristurtleDev replied:
Each of the sections in Customizing the SpriteFont touch on these points.  I think putting a new section, even small, afterwards is just repeating the information already presented.

Instead, I've added a section transition paragraph that brings this thought to the front as they are going into the Customizing the SpriteFont section

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,229 @@
+[!code-csharp[](./snippets/drawstring_full.cs)]
+
+> [!NOTE]
+> Many of these parameters (`rotation`, `origin`, `scale`, `effects`, and `layerDepth`) work exactly the same way as they do for texture rendering, as explained in [Chapter 07](../07_working_with_textures/index.md). If you need a refresher on how these parameters affect rendering, refer back to that chapter.
```
VERY nice callout !!!

AristurtleDev replied:
Thanks lol

SimonDarksideJ reviewed (changes requested):
Some feedback added

AristurtleDev committed:
Explain why SpriteFonts use a texture atlas

Added benefits for why MonoGame uses a texture atlas for fonts including:

- Cross-platform compatibility
- Consistency
- GPU Rendering
- Performance

Also includes section calrifying that text rendering involves drawing portions of the atlas for each glyph.

AristurtleDev committed:
Added admonition about Arial font

Added an Important admonition block explianing that Arial is only available on windows builts and is default default due to legacy XNA.  Included recommendation for using Roboto as an alternative.

AristurtleDev committed:
Updated Size section

Now includes information on different factors to consider when choosing a font size and different approaches to using multiple font sizes, such as multiple SpriteFont Descriptions or a single SpritFont Description of a large font and the tradeoffs between each

AristurtleDev committed:
Added not about kerning

AristurtleDev committed:
Added section transition

AristurtleDev commented:
> Some feedback added

Updated based on feedback

SimonDarksideJ commented on articles/tutorials/building_2d_games/15_working_with_spritefonts/index.md:
```diff
@@ -0,0 +1,258 @@
+
+Creating multiple SpriteFont Description files, however, can remove some of the benefits of fonts being a texture atlas since you will now have multiple atlases for each size. You'll also now have multiple assets to manage both as asset files and references in code.
+
+An alternative approach is to create a single SpriteFont Description with a larger than needed size font, then scale it down during runtime in the game. This approach allows you to maintain the single SpriteFont Description file and single texture atlas, however, the size of the texture atlas will now be larger.
```
Make into an admonition, aka Tip! to highlight

AristurtleDev replied:
Made section into Tip admonition

SimonDarksideJ reviewed (changes requested):
Looks good with the last comment above

AristurtleDev committed:
Make section a tip admoition to stand out
